src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (c) 2011-2015, jcabi.com
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met: 1) Redistributions of source code must retain the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer. 2) Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided
+ * with the distribution. 3) Neither the name of the jcabi.com nor
+ * the names of its contributors may be used to endorse or promote
+ * products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
+ * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.http.wire;
+
+import com.jcabi.http.Request;
+import com.jcabi.http.Response;
+import com.jcabi.http.Wire;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This is the base class to handle http responses with 304 state.
+ *
+ * @author Ievgen Degtiarenko (ievgen.degtiarenko@gmail.com)
+ * @version $Id$
+ * @since 2.0
+ */
+public abstract class AbstractHeaderBasedCachingWire implements Wire {
+
+    /**
+     * Cache.
+     */
+    private final transient Map<Request, Response> cache;
+
+    /**
+     * Original wire.
+     */
+    private final transient Wire origin;
+
+    /**
+     * This header will be return by server to identify content version.
+     */
+    private final transient String scvh;
+
+    /**
+     * This header will be sent by client to check if server content has
+     * been changed.
+     */
+    private final transient String cmch;
+
+    /**
+     * Ctor.
+     * @param scvh Server Response Version Header name
+     * @param cmch Client Modification Check Header name
+     * @param wire Original wire
+     */
+    AbstractHeaderBasedCachingWire(
+        final String scvh,
+        final String cmch,
+        final Wire wire
+    ) {
```

Comment:
@gmandnepr lines above do not follow http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#indentation - they take 5 lines, but they could just 2